compiler/src/iree/compiler/Codegen/Common/LinkTuningSpecsPass.cpp

```diff
@@ -61,6 +62,243 @@ static bool consumesInputOp(NamedSequenceOp op) {
+//   `_`.
+// - Otherwise, the full name is used.
+static std::string getBaseName(StringRef specName) {
+  auto pos = specName.rfind('_');
+  if (pos != StringRef::npos) {
+    StringRef potentialBase = specName.substr(0, pos);
+    unsigned suffix;
+    if (!specName.substr(pos + 1).getAsInteger(10, suffix)) {
+      return potentialBase.str(); // If valid number, return base name.
+    }
+  }
+  return specName.str(); // Otherwise, return full name.
+}
+
+static std::string
+getUniqueSpecName(StringRef specName,
+                  llvm::StringMap<unsigned> &specNameCounts) {
+  std::string specBaseName = getBaseName(specName);
+  unsigned specNameSeenCount = specNameCounts[specBaseName]++;
+
+  if (specNameSeenCount == 0)
+    return specBaseName;
+
+  return llvm::formatv("{}_{}", specBaseName, specNameSeenCount).str();
+}
+
+struct TuningSpecsToMerge {
+  SmallVector<NamedSequenceOp> namedSequenceOpsToMove;
+  // Maps a `NamedSequenceOp` to its user `ForeachMatchOp`, in which the
+  // `NamedSequenceOp` is used as a matcher or action inside a
+  // `ForeachMatchOp`.
+  llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> namedSequenceToUser;
+};
+
+// Populates the mapping of `NamedSequenceOp` to `ForeachMatchOp`
+// by processing matchers and actions inside the given `ForeachMatchOp`.
+//
+// This function iterates over all matcher and action references in
+// `foreachMatch`. If a reference is a valid `SymbolRefAttr` that resolves to a
+// `NamedSequenceOp` within the `module`, the function records the association
+// by mapping the `NamedSequenceOp` to `foreachMatch` in `namedSequenceToUser`.
+static void updateForeachMatchMappings(
+    ForeachMatchOp foreachMatch, ModuleOp module,
+    llvm::DenseMap<NamedSequenceOp, ForeachMatchOp> &namedSequenceToUser) {
+  for (auto matcher : foreachMatch.getMatchers()) {
+    if (auto matcherSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(matcher)) {
+      if (auto matcherOp =
+              SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+                  module, matcherSymRef)) {
+        namedSequenceToUser[matcherOp] = foreachMatch;
+      }
+    }
+  }
+
+  for (auto action : foreachMatch.getActions()) {
+    if (auto actionSymRef = mlir::dyn_cast<mlir::SymbolRefAttr>(action)) {
+      if (auto actionOp = SymbolTable::lookupNearestSymbolFrom<NamedSequenceOp>(
+              module, actionSymRef)) {
+        namedSequenceToUser[actionOp] = foreachMatch;
+      }
+    }
+  }
+}
+
+// Collects tuning specs (`NamedSequenceOp`s) from all inner modules within the
+// given input model, and organizes them into a `TuningSpecsToMerge` struct.
+// - `namedSequenceOpsToMove`: Contains `NamedSequenceOp`s that either:
+//     - Are not named `__kernel_config`.
+//     - Do not have the `iree_codegen.tuning_spec_entrypoint` attribute.
+// - `namedSequenceToForeachMatch`: Maps `NamedSequenceOp`s used as matchers or
+//    actions to their corresponding `ForeachMatchOp`.
```

Comment:
I think this is already covered by the comments in the struct definition. Do we need it here?